### PR TITLE
[RISCV] Add 32 bit GPR sub-register for Zfinx.

### PR DESCRIPTION
**Pull request by topperc**
This patches adds a 32 bit register class for use with Zfinx instructions.
This makes them more similar to F instructions and allows us to
only spill 32 bits.
    
I've added CodeGenOnly instructions for load/store using GPRF32 as that
gave better results than insert_subreg/extract_subreg. I'm using FSGNJ for
GPRF32 copy with Zfinx as that gave better results from MachineCopyPropagation.
    
Function arguments use this new GPRF32 register class for f32 arguments
with Zfinx. Eliminating the need to use RISCVISD::FMV* nodes.
    
This is similar to #107446 which adds a 16 bit register class.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp [hunk 1]
```diff
@@ -723,6 +772,17 @@ bool llvm::CC_RISCV_GHC(unsigned ValNo, MVT ValVT, MVT LocVT,
...
+        RISCV::X21_W, RISCV::X22_W, RISCV::X23_W, RISCV::X24_W,
+        RISCV::X25_W, RISCV::X26_W, RISCV::X27_W};
+    if (MCRegister Reg = State.AllocateReg(GPR32List)) {
+      State.addLoc(CCValAssign::getReg(ValNo, ValVT, Reg, LocVT, LocInfo));
+      return false;
+    }
+  }
+
   if ((LocVT == MVT::f32 && Subtarget.hasStdExtZfinx()) ||
       (LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:787]:
```suggestion
  if ((LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoF.td [hunk 2]
```diff
@@ -306,6 +305,19 @@ def FLW : FPLoad_r<0b010, "flw", FPR32, WriteFLD32>;
 def FSW : FPStore_r<0b010, "fsw", FPR32, WriteFST32>;
 } // Predicates = [HasStdExtF]
 
+let Predicates = [HasStdExtZfinx], isCodeGenOnly = 1 in {
+def LW_INX : Load_ri<0b010, "lw", GPRF32>, Sched<[WriteLDW, ReadMemBase]>;
+def SW_INX : Store_rri<0b010, "sw", GPRF32>,
+             Sched<[WriteSTW, ReadStoreData, ReadMemBase]>;
+
+// ADDI with GPRF16 register class to use for copy. This should not be used as
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVInstrInfoF.td:313]:
```suggestion
// ADDI with GPRF32 register class to use for copy. This should not be used as
```

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp [hunk 3]
```diff
@@ -364,11 +401,17 @@ bool llvm::CC_RISCV(unsigned ValNo, MVT ValVT, MVT LocVT,
     }
   }
 
+  if ((ValVT == MVT::f32 && Subtarget.hasStdExtZfinx())) {
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:404]:
```suggestion
  if (ValVT == MVT::f32 && Subtarget.hasStdExtZfinx()) {
```

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp [hunk 4]
```diff
@@ -723,8 +772,18 @@ bool llvm::CC_RISCV_GHC(unsigned ValNo, MVT ValVT, MVT LocVT,
...
+        RISCV::X21_W, RISCV::X22_W, RISCV::X23_W, RISCV::X24_W,
+        RISCV::X25_W, RISCV::X26_W, RISCV::X27_W};
+    if (MCRegister Reg = State.AllocateReg(GPR32List)) {
+      State.addLoc(CCValAssign::getReg(ValNo, ValVT, Reg, LocVT, LocInfo));
+      return false;
+    }
+  }
+
+  if ((LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
        Subtarget.is64Bit())) {
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:787]:
```suggestion
  if (LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
       Subtarget.is64Bit()) {
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers :)

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp [hunk 5]
```diff
@@ -156,6 +156,23 @@ static ArrayRef<MCPhysReg> getArgGPR16s(const RISCVABI::ABI ABI) {
   return ArrayRef(ArgIGPRs);
 }
 
+static ArrayRef<MCPhysReg> getArgGPR32s(const RISCVABI::ABI ABI) {
+  // The GPRs used for passing arguments in the ILP32* and LP64* ABIs, except
+  // the ILP32E ABI.
+  static const MCPhysReg ArgIGPRs[] = {RISCV::X10_W, RISCV::X11_W, RISCV::X12_W,
+                                       RISCV::X13_W, RISCV::X14_W, RISCV::X15_W,
+                                       RISCV::X16_W, RISCV::X17_W};
+  // The GPRs used for passing arguments in the ILP32E/ILP64E ABI.
```
Comment (asb) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:165]: ILP64E => LP64E

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp [hunk 6]
```diff
@@ -723,9 +772,18 @@ bool llvm::CC_RISCV_GHC(unsigned ValNo, MVT ValVT, MVT LocVT,
     }
   }
 
-  if ((LocVT == MVT::f32 && Subtarget.hasStdExtZfinx()) ||
-      (LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
-       Subtarget.is64Bit())) {
+  if (LocVT == MVT::f32 && Subtarget.hasStdExtZfinx()) {
```
Comment (asb) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:775]: Not really an issue for this review directly, but I wonder if we should be more aggressive about erroring out of the GHC calling convention for more obscure targets that are unlikely to ever need it. Is the ghc cc realistically ever going to be needed on a zfinx/zdinx system?

Change looks fine though.

**Review by asb — APPROVED** [review]
LGTM.

**Pull request merged**